modules/context/api.go

```diff
@@ -70,3 +71,29 @@ func APIContexter() macaron.Handler {
 		c.Map(ctx)
 	}
 }
+
+// Extracts the owner and repo from an APIContext, or issues an error status
+// if the operation is not applicable.
+func ExtractOwnerAndRepo(ctx *APIContext) (*models.User, *models.Repository) {
+	owner, err := models.GetUserByName(ctx.Params(":username"))
+	if err != nil {
+		if models.IsErrUserNotExist(err) {
+			ctx.Error(422, "", err)
+		} else {
+			ctx.Error(500, "GetUserByName", err)
+		}
+		return nil, nil
+	}
+
+	repo, err := models.GetRepositoryByName(owner.ID, ctx.Params(":reponame"))
+	if err != nil {
+		if models.IsErrRepoNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "GetRepositoryByName", err)
+		}
+		return nil, nil
+	}
+
+	return owner, repo
```

Comment:
Preferably, a middleware injects this into `ctx.Data["Owner"], ctx.Data["Repo"]` and returns nothing. Then it can be used as follows:
```
m.Group("/starred", context.ExtractOwnerAndRepo(), func() {
	m.Get("", user.GetMyStarredRepos)
	m.Group("/:username/:reponame", func() {
		m.Get("", user.IsStarring)
		m.Put("", user.Star)
		m.Delete("", user.Unstar)
	})
})
```

see [context.RepoRef()](https://github.com/go-gitea/gitea/blob/ddee4c8b5853d876ebd45ea4e7fe56208cd30eb5/modules/context/repo.go#L342) for reference :slightly_smiling_face: